### PR TITLE
Add solutions for Lecture 1

### PR DESCRIPTION
**Pull request by cfax**
Explicitly setting the Integer type for literals, such as 2 and 0, feels a bit overkill. Then again, if the compiler warns about doing it itself if we do not.

Use `let ... in` to make the recursive call easier to understand. Still believe `lowerCount` and `greaterCount` could be improved further.
In general `lowerAndGreater` seems a bit clunky and definitely not elegant. Not sure how to keep lines shorter than 80 chars either. Maybe to revisit.

### Solutions for Lecture {NUMBER}

cc @chshersh

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -81,7 +84,11 @@ and 1 is the smallest, and 7 - 1 = 6.
 Try to use local variables (either let-in or where) to implement this
 function.
 -}
-minmax x y z = error "TODO"
+minmax :: Integer -> Integer -> Integer -> Integer
+minmax x y z = difference
+    where difference = biggest - smallest
```
Comment (cfax) [inline src/Lecture1.hs]: In hindsight, using `difference` is a bit silly :facepalm:

File: src/Lecture1.hs [hunk 2]
```diff
@@ -54,7 +55,8 @@ Explanation: @sumOfSquares 3 4@ should be equal to @9 + 16@ and this
 is 25.
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-sumOfSquares x y = error "TODO!"
+sumOfSquares :: Integer -> Integer -> Integer
+sumOfSquares x y = x ^ (2 :: Int) + y ^ (2 :: Int)
```
Comment (chshersh) [inline src/Lecture1.hs]: This is the correct solution 👍🏻 I agree that it looks a bit ugly though 😕 

Another approach is to use `*` in this particular case to avoid ambiguity:

```suggestion
sumOfSquares x y = x * x + y * y
```

File: src/Lecture1.hs [hunk 3]
```diff
@@ -67,7 +69,8 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Integer -> Integer
+lastDigit n = mod (abs n) 10
```
Comment (chshersh) [inline src/Lecture1.hs]: Absolutely correct 👍🏻 
Alternatively, you can use the infix form of a function by putting it inside backticks :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Lecture1.hs (hunk 1 above)
Comment (chshersh) [reply src/Lecture1.hs]: I agree, probably  you can write just `biggest - smallest` in the function result itself 🙂

File: src/Lecture1.hs [hunk 4]
```diff
@@ -98,7 +105,15 @@ start position can be considered as zero (e.g. substring from the
 first character) and negative end position should result in an empty
 string.
 -}
-subString start end str = error "TODO"
+subString :: Int -> Int -> String -> String
+subString start end str = take numOfCharsToKeep $ drop start str
+    where numOfCharsToKeep = end' - start' + 1
+          start' = if (start < 0)
+                      then 0
+                      else start
```
Comment (chshersh) [inline src/Lecture1.hs]: You can use the standard function `max` here to write this slightly shorter:
```suggestion
          start' = max 0 start
```

File: src/Lecture1.hs [hunk 5]
```diff
@@ -98,7 +105,15 @@ start position can be considered as zero (e.g. substring from the
...
+          end' = if (end < 0 || end < start)
+                    then start - 1
+                    else end
```
Comment (chshersh) [inline src/Lecture1.hs]: It's a clever solution and I can see how by defining `end` lower than `start` you can achieve `numOfCharsToKeep` equal to zero.

But this code is a bit hard to follow. For clarity, I would use guards in this function. Something like this:

```haskell
subString start end str
    | end < 0     = ""
    | end < start = ""
    | otherwise   = ...
  where
    ...
```

File: src/Lecture1.hs [hunk 6]
```diff
@@ -108,7 +123,11 @@ and finds a sum of the numbers inside this string.
 
 The string contains only spaces and/or numbers.
 -}
-strSum str = error "TODO"
+strSum :: String -> Integer
+strSum str = sum ints
+    where ints = map toInt strings
+          toInt s = read s :: Integer
+          strings = words str
```
Comment (chshersh) [inline src/Lecture1.hs]: As you specify the top-level type signature for this function, the compiler would be able to figure out what type `read` function has in this case due to type inference, so you can skip specifying it in the body.

As a consequence, you can write this function as a pretty one-liner 🙃 

```suggestion
strSum str = sum (map read (words str))
```

File: src/Lecture1.hs [hunk 7]
```diff
@@ -123,4 +142,15 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 
 🕯 HINT: Use recursion to implement this function.
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: Integer -> [Integer] -> String
+lowerAndGreater n list = go n (0 :: Integer) (0 :: Integer) list
+    where go n' lower greater [] = show n' ++ " is greater than " ++ show greater ++ " elements and lower than " ++ show lower ++ " elements"
```
Comment (chshersh) [inline src/Lecture1.hs]: You can use standard function `unwords` to join a list of strings with spaces. This way it can look a bit nicer:

```suggestion
  where
    go n' lower greater [] = unwords
        [ show n'
        , "is greater than"
        , show greater
        , "elements and lower than"
        , show lower
        , "elements"
        ]
```

File: src/Lecture1.hs [hunk 8]
```diff
@@ -123,4 +142,15 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
...
+          go n' lower greater (x:rest) = let newLowerCount = lower + (lowerCount n' x)
+                                             newGreaterCount = (greater + (greaterCount n' x))
+                                          in go n' newLowerCount newGreaterCount $ rest
+          lowerCount n'' x = if (n'' < x)
+                               then 1
+                               else 0
+          greaterCount n'' x = if (n'' > x)
+                                 then 1
+                                 else 0
```
Comment (chshersh) [inline src/Lecture1.hs]: Alternatively, you could use guards (and in the future even `compare` after I explain pattern-matching) to write this without helper functions. Something like this:

```haskell
go n' lower greater (x:rest)
    | x < n' = go n' (lower + 1) greater rest
    | x < n' = go n' lower (greater + 1) rest
    | otherwise = go n' lower greater rest
```

File: src/Lecture1.hs [hunk 9]
```diff
@@ -123,4 +142,15 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 
 🕯 HINT: Use recursion to implement this function.
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: Integer -> [Integer] -> String
+lowerAndGreater n list = go n (0 :: Integer) (0 :: Integer) list
```
Comment (chshersh) [inline src/Lecture1.hs]: You don't need to pass `n` as an argument to `go` because it's in scope for the local function inside `where`. So you can just use `n` from the function definition without the need to pass it.

Also, you can specify the type of `go` explicitly and you could pass `0` without the need to specify types. And it'll be just:

```haskell
lowerAndGreater n list = go 0 0 list
```

⚠️ Spoiler alert: in the future I'll explain eta-reduction. So it's actually possible to reduce "common suffix of arguments" from both sides of `=` and the definition of this function will become even simpler:

```haskell
lowerAndGreater n = go 0 0
```

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

File: src/Lecture1.hs (hunk 3 above)
Comment (cfax) [reply src/Lecture1.hs]: Definitely  more readable!

**Comment by chshersh** [comment]
@cfax Looks great 👍🏻 I think this PR is ready for merge 🚢
